### ISSUE TITLE
Evaluate report/segment field conditions on the same related entity together

### ISSUE THREAD

**Issue by vtsykun** [oroinc/platform#738]
I am trying to make a Segment with the following conditions: "All entity that have the primary address with city LIKE %minsk%
Look like:
1) env https://demo-b2b.orocrm.com OroPlatform 2.3
2) Create 2 address
![selection_102](https://user-images.githubusercontent.com/21358010/29405474-b0ddbb08-8346-11e7-8dcd-4c80b4b555a5.png)
3) Create segment 
![selection_101](https://user-images.githubusercontent.com/21358010/29405500-c0432bc8-8346-11e7-89c2-7f07e6d86a70.png)

**Actual Result**
This entity exist in segment 
![selection_103](https://user-images.githubusercontent.com/21358010/29405648-26faf850-8347-11e7-94c9-36f54d540920.png)

**Comment by vtsykun** [comment]
In my example I want to choose entity that have an address: "city Minks is primary address", so expected result is empty datagrid, but if exist entity with have 2 address `{address1: {city:Minsk, primary: false},address2: {city:Kiev, primary: true}}` - the segment will be contains this entity

**Comment by vbyndych** [comment]
Hi @vtsykun,
Sure, I see your point. But could you do me a favor and instead of setting up 2 filters in segment:
- addresses.city contains 'Minsk'
- addresses.primary is Yes
Use one filter:
- primary_addr.city contains 'Minsk'

If you do so you will get an empty list, which as I understand is correct for your example. Please let me know your results.

P.S. I just don't have Minsk in my test data, so I've used another city - Long.

**Comment by mbessolov** [comment]
@vtsykun  This is the expected behavior in OroPlatform.
Each of the filters is applied separately and produces its own subset of the main entities. Those subsets then are merged together using the specified logical operations.
In your example, we first retrieve a subset of Contacts, who have one or more addresses with Minsk as the city. Then, we retrieve the second subset of Contacts, who have some address marked as primary. And then we merge those subsets, so the result is the subset of Contacts, that have some Minsk address(es) and some address marked as primary (which may be any address, not necessarily the Minsk address).

There is no way to specify in the filter editor, that those field conditions should be applied on the related entity (address) together. This feature (ability to specify that some field conditions are to be applied on a related entity together) is on the roadmap, but there are no specific plans at this time as to in which release it will be available.

Meanwhile, in order to build the desired report  you need to start with a different entity. In your example it should be the contact address entity. Specify the field conditions on the contact address entity, and then group contact addresses by contact field to eliminate duplicate contacts, and add the contact field as report columns.

**Comment by vtsykun** [comment]
Ok, I understood, thanks

**Comment by manowark** [comment]
I'm closing this issue as it seems clarified. Please, feel free to reopen it if you have any questions.

**Comment by vtsykun** [comment]
Hi @manowark 
I noticed that in old OroPlatform (v1.9.2) version before merge https://github.com/oroinc/platform/search?q=BAP-11180&type=Commits&utf8=%E2%9C%93 (BAP-11180: Filter does not work if two or more conditions are applied…) it was worked exactly as we want. So I reverted this changes and get expected result.
For example.
![selection_179](https://user-images.githubusercontent.com/21358010/30316722-bfab5036-97b0-11e7-899c-67416202c0ef.png)

**Before merge BAP-11180**
```
SELECT 
  o2_.last_name AS last_name_0, 
  o2_.id AS id_1, 
  o2_.organization_id AS sclr_2, 
  o2_.businessUnit_id AS sclr_3 
FROM 
  orocrm_contact o2_ 
WHERE 
  o2_.id IN (
    SELECT 
      o3_.id 
    FROM 
      orocrm_contact o3_ 
      LEFT JOIN orocrm_contact_email o4_ ON o3_.id = o4_.owner_id 
      LEFT JOIN orocrm_contact_phone o5_ ON o3_.id = o5_.owner_id 
      LEFT JOIN oro_enum_contact_phone_type o0_ ON o5_.phoneType_id = o0_.id 
      LEFT JOIN oro_enum_value_trans t1_ ON t1_.locale = 'es_ES' 
      AND t1_.field = 'name' 
      AND t1_.object_class = 'Extend\Entity\EV_Contact_Phone_Type' 
      AND t1_.foreign_key = o0_.id 
    WHERE 
      o4_.is_primary = true 
      AND o4_.email LIKE ? 
      AND (
        o5_.is_primary = true 
        AND o5_.phoneType_id = ?
      ) 
      AND o3_.last_name LIKE ? 
      AND o3_.organization_id = 1
  ) 
  AND o2_.organization_id = 1 
ORDER BY 
  o2_.id ASC 
LIMIT 
  25
```
**After merge BAP-11180**
```
SELECT 
  o10_.last_name AS last_name_0, 
  o10_.id AS id_1, 
  o10_.organization_id AS sclr_2, 
  o10_.businessUnit_id AS sclr_3 
FROM 
  orocrm_contact o10_ 
WHERE 
  o10_.id IN (
    SELECT 
      o11_.id 
    FROM 
      orocrm_contact o11_ 
      LEFT JOIN orocrm_contact_email o12_ ON o11_.id = o12_.owner_id 
      LEFT JOIN orocrm_contact_phone o13_ ON o11_.id = o13_.owner_id 
      LEFT JOIN oro_enum_contact_phone_type o0_ ON o13_.phoneType_id = o0_.id 
      LEFT JOIN oro_enum_value_trans t1_ ON t1_.locale = 'es_ES' 
      AND t1_.field = 'name' 
      AND t1_.object_class = 'Extend\Entity\EV_Contact_Phone_Type' 
      AND t1_.foreign_key = o0_.id 
    WHERE 
      (
        EXISTS (
          SELECT 
            o14_.id 
          FROM 
            orocrm_contact o14_ 
            LEFT JOIN orocrm_contact_email o15_ ON o14_.id = o15_.owner_id 
            LEFT JOIN orocrm_contact_phone o16_ ON o14_.id = o16_.owner_id 
            LEFT JOIN oro_enum_contact_phone_type o2_ ON o16_.phoneType_id = o2_.id 
            LEFT JOIN oro_enum_value_trans t3_ ON t3_.locale = 'es_ES' 
            AND t3_.field = 'name' 
            AND t3_.object_class = 'Extend\Entity\EV_Contact_Phone_Type' 
            AND t3_.foreign_key = o2_.id 
          WHERE 
            o15_.is_primary = true 
            AND o14_.id = o11_.id
        )
      ) 
      AND (
        EXISTS (
          SELECT 
            o17_.id 
          FROM 
            orocrm_contact o17_ 
            LEFT JOIN orocrm_contact_email o18_ ON o17_.id = o18_.owner_id 
            LEFT JOIN orocrm_contact_phone o19_ ON o17_.id = o19_.owner_id 
            LEFT JOIN oro_enum_contact_phone_type o4_ ON o19_.phoneType_id = o4_.id 
            LEFT JOIN oro_enum_value_trans t5_ ON t5_.locale = 'es_ES' 
            AND t5_.field = 'name' 
            AND t5_.object_class = 'Extend\Entity\EV_Contact_Phone_Type' 
            AND t5_.foreign_key = o4_.id 
          WHERE 
            LOWER(o18_.email) LIKE LOWER(?) 
            AND o17_.id = o11_.id
        )
      ) 
      AND (
        (
          EXISTS (
            SELECT 
              o20_.id 
            FROM 
              orocrm_contact o20_ 
              LEFT JOIN orocrm_contact_email o21_ ON o20_.id = o21_.owner_id 
              LEFT JOIN orocrm_contact_phone o22_ ON o20_.id = o22_.owner_id 
              LEFT JOIN oro_enum_contact_phone_type o6_ ON o22_.phoneType_id = o6_.id 
              LEFT JOIN oro_enum_value_trans t7_ ON t7_.locale = 'es_ES' 
              AND t7_.field = 'name' 
              AND t7_.object_class = 'Extend\Entity\EV_Contact_Phone_Type' 
              AND t7_.foreign_key = o6_.id 
            WHERE 
              o22_.is_primary = true 
              AND o20_.id = o11_.id
          )
        ) 
        AND (
          EXISTS (
            SELECT 
              o23_.id 
            FROM 
              orocrm_contact o23_ 
              LEFT JOIN orocrm_contact_email o24_ ON o23_.id = o24_.owner_id 
              LEFT JOIN orocrm_contact_phone o25_ ON o23_.id = o25_.owner_id 
              LEFT JOIN oro_enum_contact_phone_type o8_ ON o25_.phoneType_id = o8_.id 
              LEFT JOIN oro_enum_value_trans t9_ ON t9_.locale = 'es_ES' 
              AND t9_.field = 'name' 
              AND t9_.object_class = 'Extend\Entity\EV_Contact_Phone_Type' 
              AND t9_.foreign_key = o8_.id 
            WHERE 
              o25_.phoneType_id = ? 
              AND o23_.id = o11_.id
          )
        )
      ) 
      AND LOWER(o11_.last_name) LIKE LOWER(?) 
      AND o11_.organization_id = 1
  ) 
  AND o10_.organization_id = 1 
ORDER BY 
  o10_.id ASC 
LIMIT 
  25
```
My  question here would be: why did you swtched to new "join strategy"? - This is major BC in patch realese and major performance bug. Maybe there's a flaw in old strategy that we're not seeing yet and it doesn't work.

\cc @mbessolov

**Comment by sergeyz** [comment]
@vtsykun Thanks for report, INTERNAL ID: BAP-12268

**Comment by anyt** [comment]
Implemented in 3.1.7 version, see https://github.com/oroinc/platform/commit/70071776ed70ee4eebe92ca894eddbde35173a99.
